File: web/lib/engagement.ts
export interface NetworkEngagementTrace {
  yeo_key: string;
  headline: string;
  tagline: string;
  parcel_count: number;
  raw: number[];
  zscore: number[];
  derivative: number[];
}

export interface NetworkSummary {
  mean_z: number;
  peak_z: number;
  peak_tr: number;
  active_fraction: number;
}

export interface DominantSegment {
  net: string;
  start_tr: number;
  end_tr: number;
  duration_s: number;
}

export interface EpochTemplate {
  id: string;
  label: string;
  hypothesis: string;
  score: number;
  caveat: string;
}

export interface EpochSegment {
  net: string;
  start_tr: number;
  end_tr: number;
  duration_s: number;
  template?: EpochTemplate;
}

export interface AcousticFeatureTrace {
  label: string;
  description: string;
  raw: number[];
  zscore: number[];
}

export interface AcousticFeaturesData {
  version: number;
  fps: number;
  n_trs: number;
  source: string;
  feature_order: string[];
  features: Record<string, AcousticFeatureTrace>;
  disclaimer: string;
}

export interface EngagementData {
  version: number;
  fps: number;
  n_trs: number;
  atlas: {
    name: string;
    n_parcels: number;
    yeo_networks: number;
  };
  inference_mode: string;
  preprocessing: {
    network_aggregate: string;
    zscore: string;
    skip_trs: number;
    macro_rollup?: boolean;
  };
  networks: Record<string, NetworkEngagementTrace>;
  subnetworks?: Record<string, NetworkEngagementTrace>;
  sub_summaries?: Record<string, NetworkSummary>;
  summaries: Record<string, NetworkSummary>;
  derived: {
    dominant_network_tr: string[];
    dominant_segments?: DominantSegment[];
    epoch_segments?: EpochSegment[];
    coupling: Record<
      string,
      { networks: string[]; window_s: number; series: number[] }
    >;
    salience_events: {
      threshold_z_derivative: number;
      trs: number[];
    };
    deep_dive_subnets?: Record<string, string[]>;
  };
  disclaimer: string;
}

/** Yeo-7 display order — matches backend YEO_NETWORK_ORDER. */
export const ENGAGEMENT_NETWORK_ORDER = [
  "Cont",
  "SalVentAttn",
  "DorsAttn",
  "Default",
  "Limbic",
  "SomMot",
  "Vis",
] as const;

/** Connectome Workbench / Schaefer Yeo-7 colors (CSS rgb). */
export const ENGAGEMENT_COLORS: Record<string, string> = {
  Vis: "rgb(120, 18, 134)",
  SomMot: "rgb(70, 130, 180)",
  DorsAttn: "rgb(0, 118, 14)",
  SalVentAttn: "rgb(230, 148, 34)",
  Limbic: "rgb(220, 0, 115)",
  Cont: "rgb(120, 94, 224)",
  Default: "rgb(255, 0, 0)",
};

export function engagementNetworksInOrder(
  data: EngagementData
): NetworkEngagementTrace[] {
  return ENGAGEMENT_NETWORK_ORDER.map((key) => data.networks[key]).filter(Boolean);
}

/** Format seconds as m:ss for segment labels. */
export function formatSegmentTime(seconds: number): string {
  const m = Math.floor(seconds / 60);
  const s = seconds % 60;
  return `${m}:${String(s).padStart(2, "0")}`;
}

/** Parse dominant-network runs from per-TR labels (fallback if segments absent). */
export function parseDominantSegments(
  dominantNetworkTr: string[]
): DominantSegment[] {
  if (dominantNetworkTr.length === 0) return [];

  const segments: DominantSegment[] = [];
  let start = 0;
  let current = dominantNetworkTr[0];

  for (let i = 1; i < dominantNetworkTr.length; i++) {
    if (dominantNetworkTr[i] !== current) {
      segments.push({
        net: current,
        start_tr: start,
        end_tr: i - 1,
        duration_s: i - start,
      });
      start = i;
      current = dominantNetworkTr[i];
    }
  }

  segments.push({
    net: current,
    start_tr: start,
    end_tr: dominantNetworkTr.length - 1,
    duration_s: dominantNetworkTr.length - start,
  });

  return segments;
}

export function getDominantSegments(data: EngagementData): DominantSegment[] {
  if (data.derived.dominant_segments?.length) {
    return data.derived.dominant_segments;
  }
  return parseDominantSegments(data.derived.dominant_network_tr);
}

export function networkFractions(
  dominantNetworkTr: string[]
): Record<string, number> {
  if (dominantNetworkTr.length === 0) return {};
  const counts: Record<string, number> = {};
  for (const net of dominantNetworkTr) {
    counts[net] = (counts[net] ?? 0) + 1;
  }
  const n = dominantNetworkTr.length;
  return Object.fromEntries(
    Object.entries(counts).map(([net, c]) => [net, c / n])
  );
}

export interface NetworkEngagementDescription {
  /** Yeo-7 network name for readers who want the neuroscience label. */
  yeoLabel: string;
  /** Short paragraphs shown when a row is expanded. */
  paragraphs: string[];
}

/** Expanded copy for each headline metric (frontend-only). */
export const ENGAGEMENT_DESCRIPTIONS: Record<
  (typeof ENGAGEMENT_NETWORK_ORDER)[number],
  NetworkEngagementDescription
> = {
  Cont: {
    yeoLabel: "Control network",
    paragraphs: [
      "The control network coordinates executive processes: holding goals in mind, resolving competition between responses, and regulating other networks. It is often engaged when a task requires sustained effort or flexible updating.",
      "In music, Focus may rise during dense or structurally complex passages—polyphony, rapid lyrical delivery, or sections that invite active parsing rather than background listening.",
      "Each point is the mean predicted BOLD-like signal across Schaefer parcels mapped to Yeo-7 Control, z-scored within this clip so you can compare moments relative to the track itself.",
    ],
  },
  SalVentAttn: {
    yeoLabel: "Salience / ventral attention network",
    paragraphs: [
      "The salience network detects events that stand out from the ongoing stream and helps reorient attention. It is central to bottom-up surprise, novelty, and salient sensory change.",
      "Surprise often spikes at drops, entry of a new timbre, sudden dynamic shifts, or boundary moments between sections. Sharp upward moves in this trace are also used to flag salience events elsewhere in the panel.",
      "This trace aggregates parcels in Yeo-7 Salience/Ventral Attention and is z-scored within the clip.",
    ],
  },
  DorsAttn: {
    yeoLabel: "Dorsal attention network",
    paragraphs: [
      "The dorsal attention network supports top-down, goal-directed attention: selecting what to follow in a noisy environment and maintaining that focus over time.",
      "Tracking may stay elevated when a lead melody, rhythm, or motif holds your attention across bars—steady grooves, repeating hooks, or passages you are deliberately following.",
      "Values come from Yeo-7 Dorsal Attention parcels, averaged and z-scored within the clip.",
    ],
  },
  Default: {
    yeoLabel: "Default mode network",
    paragraphs: [
      "The default mode network is associated with internally directed thought: autobiographical memory, narrative construction, and self-relevant interpretation. It often quiets during demanding external tasks but can re-engage during rich, meaningful experience.",
      "Resonance may increase when music feels personal, story-like, or emotionally reflective—lyrical meaning, nostalgic tone, or passages that invite imagination beyond the sound itself.",
      "This trace reflects Yeo-7 Default network parcels, z-scored within the clip.",
    ],
  },
  Limbic: {
    yeoLabel: "Limbic network",
    paragraphs: [
      "Limbic-related cortex supports affective evaluation—pleasure, tension, reward anticipation, and emotional coloring of experience. It interacts closely with memory and motivational systems.",
      "Feeling may track harmonic warmth, vocal expressiveness, lyrical valence, or build-and-release dynamics that carry emotional weight rather than purely sensory surprise.",
      "Aggregated from Yeo-7 Limbic parcels and z-scored within the clip.",
    ],
  },
  SomMot: {
    yeoLabel: "Somatomotor network",
    paragraphs: [
      "The somatomotor network spans primary and supplementary motor regions plus related somatosensory areas. It is engaged by movement planning, rhythmic timing, and sensorimotor coupling.",
      "Pulse often follows beat strength, syncopation, and groove—sections that invite foot-tapping, dance, or internal entrainment to the rhythm.",
      "Derived from Yeo-7 Somatomotor parcels, averaged and z-scored within the clip.",
    ],
  },
  Vis: {
    yeoLabel: "Visual network",
    paragraphs: [
      "The visual network processes visual input and supports internally generated visual imagery. Auditory stimuli can co-engage visual cortex when listening evokes scenes, motion, or vivid timbral “color.”",
      "Imagery may rise with cinematic production, spatial mixing, or passages that feel visually evocative—even without any video present.",
      "This trace averages Yeo-7 Visual parcels and is z-scored within the clip.",
    ],
  },
};

/** Key Yeo-17 subnetworks shown under each macro headline row. */
export const MACRO_DEEP_DIVE_SUBNETS: Record<
  (typeof ENGAGEMENT_NETWORK_ORDER)[number],
  readonly string[]
> = {
  Cont: ["ContA", "ContB", "ContC"],
  SalVentAttn: ["SalVentAttnA", "SalVentAttnB"],
  DorsAttn: ["DorsAttnA", "DorsAttnB"],
  Default: ["DefaultA", "DefaultB", "DefaultC", "TempPar"],
  Limbic: ["LimbicA", "LimbicB"],
  SomMot: ["SomMotA", "SomMotB"],
  Vis: ["VisCent", "VisPeri"],
};

export const SUBNETWORK_COLORS: Record<string, string> = {
  VisCent: "rgb(120, 18, 134)",
  VisPeri: "rgb(160, 60, 170)",
  SomMotA: "rgb(70, 130, 180)",
  SomMotB: "rgb(110, 160, 205)",
  DorsAttnA: "rgb(0, 118, 14)",
  DorsAttnB: "rgb(40, 160, 50)",
  SalVentAttnA: "rgb(230, 148, 34)",
  SalVentAttnB: "rgb(255, 190, 90)",
  LimbicA: "rgb(220, 0, 115)",
  LimbicB: "rgb(255, 80, 150)",
  ContA: "rgb(120, 94, 224)",
  ContB: "rgb(150, 120, 235)",
  ContC: "rgb(180, 150, 245)",
  DefaultA: "rgb(255, 0, 0)",
  DefaultB: "rgb(255, 80, 80)",
  DefaultC: "rgb(255, 130, 130)",
  TempPar: "rgb(200, 40, 40)",
};

export const SUBNETWORK_DESCRIPTIONS: Record<string, NetworkEngagementDescription> = {
  SalVentAttnA: {
    yeoLabel: "Salience A — anterior insula / cingulate",
    paragraphs: [
      "SalVentAttnA includes anterior insula and mid-cingulate regions central to interoceptive salience and bottom-up event detection.",
      "In music, this subnetwork is the best cortical proxy for sudden orienting — drops, boundary accents, and timbral shocks (when not averaged away into the macro salience trace).",
    ],
  },
  SalVentAttnB: {
    yeoLabel: "Salience B — ventral attention / TPJ",
    paragraphs: [
      "SalVentAttnB overlaps ventral attention and temporoparietal junction — reorienting after an event rather than the initial insula spike.",
    ],
  },
  DefaultA: {
    yeoLabel: "Default A — medial prefrontal core",
    paragraphs: [
      "DefaultA is the medial prefrontal hub of the DMN — self-relevant interpretation and narrative framing during listening.",
    ],
  },
  DefaultB: {
    yeoLabel: "Default B — temporal / angular DMN",
    paragraphs: [
      "DefaultB spans lateral temporal and angular regions tied to semantic association and episodic context.",
    ],
  },
  DefaultC: {
    yeoLabel: "Default C — posterior DMN",
    paragraphs: [
      "DefaultC is a posterior DMN hub — integration of autobiographical and scene-like imagery while music unfolds.",
    ],
  },
  TempPar: {
    yeoLabel: "Temporal parietal — language / narrative",
    paragraphs: [
      "TempPar bridges temporal and parietal language networks. It often co-elevates with DefaultA when lyrics or semantic narrative is foregrounded.",
    ],
  },
  ContA: {
    yeoLabel: "Control A — frontoparietal",
    paragraphs: [
      "ContA is the classic frontoparietal control system — sustained effort when parsing dense musical structure.",
    ],
  },
  ContB: {
    yeoLabel: "Control B — cingulo-insular control",
    paragraphs: [
      "ContB links mid-cingulate and insular control — flexible updating when the musical stream shifts.",
    ],
  },
  ContC: {
    yeoLabel: "Control C — posterior control",
    paragraphs: [
      "ContC supports posterior control processes — maintaining task sets across longer passages.",
    ],
  },
  DorsAttnA: {
    yeoLabel: "Dorsal attention A",
    paragraphs: ["Frontoparietal top-down selection — following a lead line or motif."],
  },
  DorsAttnB: {
    yeoLabel: "Dorsal attention B",
    paragraphs: ["Posterior dorsal attention — spatial / spectral tracking of ongoing texture."],
  },
  SomMotA: {
    yeoLabel: "Somatomotor A",
    paragraphs: ["Primary motor and premotor — beat entrainment and movement planning."],
  },
  SomMotB: {
    yeoLabel: "Somatomotor B",
    paragraphs: ["Extended somatomotor — secondary rhythmic coupling."],
  },
  LimbicA: {
    yeoLabel: "Limbic A — orbitofrontal / temporal pole",
    paragraphs: ["Affective evaluation and reward tone — warmth, tension, valence."],
  },
  LimbicB: {
    yeoLabel: "Limbic B — cingulate / parahippocampal",
    paragraphs: ["Emotional memory coloring and anticipatory affect."],
  },
  VisCent: {
    yeoLabel: "Central visual",
    paragraphs: ["Early visual cortex — vivid timbral color and centrally imagined scenes."],
  },
  VisPeri: {
    yeoLabel: "Peripheral visual",
    paragraphs: ["Peripheral visual field — spatial / panoramic imagery from production."],
  },
};

export const ACOUSTIC_OVERLAY_COLORS: Record<string, string> = {
  rms: "rgb(255, 210, 80)",
  spectral_centroid: "rgb(120, 200, 255)",
  spectral_flux: "rgb(180, 255, 140)",
  onset_strength: "rgb(255, 120, 160)",
};

export function getEpochSegments(data: EngagementData): EpochSegment[] {
  if (data.derived.epoch_segments?.length) {
    return data.derived.epoch_segments;
  }
  return getDominantSegments(data).map((seg) => ({ ...seg }));
}

export function epochTemplateForSegment(
  data: EngagementData,
  segment: DominantSegment
): EpochTemplate | undefined {
  const match = getEpochSegments(data).find(
    (s) =>
      s.net === segment.net &&
      s.start_tr === segment.start_tr &&
      s.end_tr === segment.end_tr
  );
  return match?.template;
}

export function deepDiveSubnets(
  data: EngagementData,
  macro: (typeof ENGAGEMENT_NETWORK_ORDER)[number]
): string[] {
  const fromExport = data.derived.deep_dive_subnets?.[macro];
  if (fromExport?.length) return fromExport;
  return [...MACRO_DEEP_DIVE_SUBNETS[macro]];
}

File: src/nerve/analysis/engagement.py
"""Yeo network engagement traces for web export."""

from __future__ import annotations

from typing import Any

import numpy as np

from nerve.parcellation.schaefer import (
    MACRO_DEEP_DIVE_SUBNETS,
    YEO_17_NETWORK_ORDER,
    YEO_17_TO_MACRO,
    YEO_NETWORK_ORDER,
    rollup_to_macro,
)

DISCLAIMER = (
    "Predicted group-average cortical engagement (in-silico). "
    "Exploratory; not clinical or individual inference."
)

NETWORK_META: dict[str, dict[str, str]] = {
    "Cont": {
        "headline": "Focus",
        "tagline": "Sustained active processing",
    },
    "SalVentAttn": {
        "headline": "Surprise",
        "tagline": "Bottom-up salient events",
    },
    "DorsAttn": {
        "headline": "Tracking",
        "tagline": "Sustained top-down attention",
    },
    "Default": {
        "headline": "Resonance",
        "tagline": "Internal, self-relevant processing",
    },
    "Limbic": {
        "headline": "Feeling",
        "tagline": "Affective and evaluative tone",
    },
    "SomMot": {
        "headline": "Pulse",
        "tagline": "Rhythm and motor-related coupling",
    },
    "Vis": {
        "headline": "Imagery",
        "tagline": "Visual sensory engagement",
    },
}

SUBNETWORK_META: dict[str, dict[str, str]] = {
    "VisCent": {"headline": "Central visual", "tagline": "Early visual cortex"},
    "VisPeri": {"headline": "Peripheral visual", "tagline": "Peripheral visual field"},
    "SomMotA": {"headline": "Motor A", "tagline": "Primary motor / premotor"},
    "SomMotB": {"headline": "Motor B", "tagline": "Secondary somatomotor"},
    "DorsAttnA": {"headline": "Dorsal attention A", "tagline": "Frontoparietal top-down"},
    "DorsAttnB": {"headline": "Dorsal attention B", "tagline": "Posterior attention"},
    "SalVentAttnA": {"headline": "Salience A (insula)", "tagline": "Anterior salience / insula"},
    "SalVentAttnB": {"headline": "Salience B", "tagline": "Ventral attention / TPJ"},
    "LimbicA": {"headline": "Limbic A", "tagline": "Orbitofrontal / temporal pole"},
    "LimbicB": {"headline": "Limbic B", "tagline": "Cingulate / parahippocampal"},
    "ContA": {"headline": "Control A", "tagline": "Frontoparietal control"},
    "ContB": {"headline": "Control B", "tagline": "Mid-cingulo-insular"},
    "ContC": {"headline": "Control C", "tagline": "Posterior control"},
    "DefaultA": {"headline": "Default A", "tagline": "Medial prefrontal core"},
    "DefaultB": {"headline": "Default B", "tagline": "Temporal / angular DMN"},
    "DefaultC": {"headline": "Default C", "tagline": "Posterior DMN hub"},
    "TempPar": {"headline": "Temporal parietal", "tagline": "Language / narrative integration"},
}

EPOCH_TEMPLATES: tuple[dict[str, Any], ...] = (
    {
        "id": "decoupling",
        "label": "Decoupling bridge",
        "hypothesis": (
            "Default-mode engagement rises while control dips — exploratory pattern "
            "associated with internally directed, less effortful listening."
        ),
    },
    {
        "id": "event_drop",
        "label": "Event / drop",
        "hypothesis": (
            "Salience (especially insula-linked SalVentAttnA) spikes — pattern "
            "consistent with a bottom-up orienting response to a salient acoustic event."
        ),
    },
    {
        "id": "groove",
        "label": "Groove / entrainment",
        "hypothesis": (
            "Somatomotor and limbic co-elevation — exploratory pattern linked to "
            "rhythmic motor coupling and affective groove in music cognition literature."
        ),
    },
    {
        "id": "active_parsing",
        "label": "Active parsing",
        "hypothesis": (
            "Control and dorsal attention co-elevation — pattern consistent with "
            "effortful tracking of complex or dense musical structure."
        ),
    },
    {
        "id": "narrative",
        "label": "Narrative / lyrical",
        "hypothesis": (
            "Temporal-parietal and default-A co-elevation — exploratory pattern "
            "associated with semantic / narrative processing during music with lyrics."
        ),
    },
)

ACTIVE_Z_THRESHOLD = 0.5
SALIENCE_DERIVATIVE_THRESHOLD = 1.5
COUPLING_WINDOW_TRS = 8
EPOCH_Z_THRESHOLD = 0.35

COUPLING_PAIRS: tuple[tuple[str, str], ...] = (
    ("SomMot", "Limbic"),
    ("Cont", "Default"),
    ("SalVentAttn", "SomMot"),
)


def _zscore(series: np.ndarray, eps: float = 1e-8) -> np.ndarray:
    mean = float(series.mean())
    std = float(series.std())
    if std < eps:
        return np.zeros_like(series)
    return (series - mean) / (std + eps)


def _derivative(z: np.ndarray) -> np.ndarray:
    out = np.zeros_like(z)
    if z.size > 1:
        out[1:] = z[1:] - z[:-1]
    return out


def _rolling_correlation(a: np.ndarray, b: np.ndarray, window: int) -> np.ndarray:
    n = a.size
    out = np.zeros(n, dtype=np.float64)
    if n < 2 or window < 2:
        return out
    w = min(window, n)
    for t in range(n):
        start = max(0, t - w + 1)
        chunk_a = a[start : t + 1]
        chunk_b = b[start : t + 1]
        if chunk_a.size < 2:
            continue
        sa = chunk_a.std()
        sb = chunk_b.std()
        if sa < 1e-8 or sb < 1e-8:
            continue
        out[t] = float(np.corrcoef(chunk_a, chunk_b)[0, 1])
    return out


def _network_indices(networks: list[str], order: list[str]) -> dict[str, list[int]]:
    grouped: dict[str, list[int]] = {n: [] for n in order}
    for idx, net in enumerate(networks):
        if net in grouped:
            grouped[net].append(idx)
    return grouped


def _macro_indices(networks: list[str]) -> dict[str, list[int]]:
    grouped: dict[str, list[int]] = {n: [] for n in YEO_NETWORK_ORDER}
    for idx, net in enumerate(networks):
        macro = rollup_to_macro(net)
        if macro in grouped:
            grouped[macro].append(idx)
    return grouped


def _align_networks(networks: list[str], n_parcels: int) -> list[str]:
    """Drop Background label row when labels outnumber parcel rows."""
    if len(networks) == n_parcels + 1 and networks[0] in ("Unknown", ""):
        return networks[1 : n_parcels + 1]
    if len(networks) > n_parcels:
        return networks[:n_parcels]
    if len(networks) < n_parcels:
        raise ValueError(
            f"networks length {len(networks)} != n_parcels {n_parcels}"
        )
    return networks


def compute_network_raw_traces(
    parcel_data: np.ndarray,
    networks: list[str],
    *,
    macro: bool = True,
) -> dict[str, np.ndarray]:
    """
    Aggregate parcel time series to Yeo network means.

    parcel_data: (n_parcels, T)
    macro=True → Yeo-7 rollup; macro=False → native subnetwork labels.
    """
    if parcel_data.ndim != 2:
        raise ValueError(f"expected (n_parcels, T), got {parcel_data.shape}")

    networks = _align_networks(networks, parcel_data.shape[0])
    order = YEO_NETWORK_ORDER if macro else _subnetwork_order(networks)
    grouped = _macro_indices(networks) if macro else _network_indices(networks, order)

    traces: dict[str, np.ndarray] = {}
    for net in order:
        idxs = grouped[net]
        if not idxs:
            traces[net] = np.zeros(parcel_data.shape[1], dtype=np.float64)
        else:
            traces[net] = parcel_data[idxs, :].mean(axis=0)
    return traces


def _subnetwork_order(networks: list[str]) -> list[str]:
    present = set(networks)
    if present <= set(YEO_NETWORK_ORDER):
        return list(YEO_NETWORK_ORDER)
    return [n for n in YEO_17_NETWORK_ORDER if n in present]


def _has_subnetworks(networks: list[str]) -> bool:
    return any(n in YEO_17_TO_MACRO for n in networks)


def parse_dominant_segments(dominant_network_tr: list[str]) -> list[dict[str, int | str]]:
    """Contiguous runs where each network wins the per-TR z-score argmax."""
    if not dominant_network_tr:
        return []

    segments: list[dict[str, int | str]] = []
    start = 0
    current = dominant_network_tr[0]

    for i in range(1, len(dominant_network_tr)):
        if dominant_network_tr[i] != current:
            segments.append(
                {
                    "net": current,
                    "start_tr": start,
                    "end_tr": i - 1,
                    "duration_s": i - start,
                }
            )
            start = i
            current = dominant_network_tr[i]

    segments.append(
        {
            "net": current,
            "start_tr": start,
            "end_tr": len(dominant_network_tr) - 1,
            "duration_s": len(dominant_network_tr) - start,
        }
    )
    return segments


def summarize_trace(z: np.ndarray) -> dict[str, float | int]:
    peak_tr = int(np.argmax(z)) if z.size else 0
    return {
        "mean_z": float(z.mean()) if z.size else 0.0,
        "peak_z": float(z.max()) if z.size else 0.0,
        "peak_tr": peak_tr,
        "active_fraction": float(np.mean(z > ACTIVE_Z_THRESHOLD)) if z.size else 0.0,
    }


def _segment_mean(z: np.ndarray, start: int, end: int) -> float:
    if z.size == 0:
        return 0.0
    return float(z[start : end + 1].mean())


def classify_epoch_template(
    segment: dict[str, int | str],
    macro_z: dict[str, np.ndarray],
    sub_z: dict[str, np.ndarray] | None = None,
) -> dict[str, Any] | None:
    """Score literature-inspired epoch templates for a dominant segment."""
    start = int(segment["start_tr"])
    end = int(segment["end_tr"])

    def m(key: str, traces: dict[str, np.ndarray]) -> float:
        arr = traces.get(key)
        if arr is None:
            return 0.0
        return _segment_mean(arr, start, end)

    scores: dict[str, float] = {}

    default_z = m("Default", macro_z)
    cont_z = m("Cont", macro_z)
    scores["decoupling"] = max(0.0, default_z - cont_z)

    sal_a = m("SalVentAttnA", sub_z) if sub_z else m("SalVentAttn", macro_z)
    scores["event_drop"] = sal_a

    scores["groove"] = min(m("SomMot", macro_z), m("Limbic", macro_z))

    cont_a = m("ContA", sub_z) if sub_z else cont_z
    dors_a = m("DorsAttnA", sub_z) if sub_z else m("DorsAttn", macro_z)
    scores["active_parsing"] = min(cont_a, dors_a)

    temp_par = m("TempPar", sub_z) if sub_z else 0.0
    default_a = m("DefaultA", sub_z) if sub_z else default_z
    scores["narrative"] = min(temp_par, default_a) if sub_z else 0.0

    best_id = max(scores, key=scores.get)
    best_score = scores[best_id]
    if best_score < EPOCH_Z_THRESHOLD:
        return None

    template = next(t for t in EPOCH_TEMPLATES if t["id"] == best_id)
    return {
        "id": template["id"],
        "label": template["label"],
        "hypothesis": template["hypothesis"],
        "score": round(best_score, 3),
        "caveat": (
            "Exploratory in-silico hypothesis from predicted group-average BOLD; "
            "not validated fMRI inference."
        ),
    }


def classify_segment_epoch_templates(
    segments: list[dict[str, int | str]],
    macro_z: dict[str, np.ndarray],
    sub_z: dict[str, np.ndarray] | None = None,
) -> list[dict[str, Any]]:
    out: list[dict[str, Any]] = []
    for seg in segments:
        template = classify_epoch_template(seg, macro_z, sub_z)
        entry: dict[str, Any] = {
            "net": seg["net"],
            "start_tr": seg["start_tr"],
            "end_tr": seg["end_tr"],
            "duration_s": seg["duration_s"],
        }
        if template:
            entry["template"] = template
        out.append(entry)
    return out


def _build_network_docs(
    raw_traces: dict[str, np.ndarray],
    grouped: dict[str, list[int]],
    meta: dict[str, dict[str, str]],
) -> tuple[dict[str, Any], dict[str, Any], dict[str, np.ndarray], dict[str, np.ndarray]]:
    z_traces: dict[str, np.ndarray] = {}
    d_traces: dict[str, np.ndarray] = {}
    network_docs: dict[str, Any] = {}
    summaries: dict[str, Any] = {}

    for net, raw in raw_traces.items():
        z = _zscore(raw)
        d = _derivative(z)
        z_traces[net] = z
        d_traces[net] = d
        info = meta.get(net, {"headline": net, "tagline": ""})
        network_docs[net] = {
            "yeo_key": net,
            "headline": info["headline"],
            "tagline": info["tagline"],
            "parcel_count": len(grouped.get(net, [])),
            "raw": raw.tolist(),
            "zscore": z.tolist(),
            "derivative": d.tolist(),
        }
        summaries[net] = summarize_trace(z)

    return network_docs, summaries, z_traces, d_traces


def compute_engagement(
    parcel_data: np.ndarray,
    networks: list[str],
    *,
    n_parcels: int,
    yeo_networks: int = 17,
    inference_mode: str = "acoustic_only",
    fps: int = 1,
    skip_trs: int = 0,
) -> dict[str, Any]:
    """Build engagement.json document from (n_parcels, T) parcel matrix."""
    networks = _align_networks(networks, parcel_data.shape[0])
    n_trs = int(parcel_data.shape[1])
    use_sub = _has_subnetworks(networks)

    macro_raw = compute_network_raw_traces(parcel_data, networks, macro=True)
    macro_grouped = _macro_indices(networks)
    network_docs, summaries, z_traces, d_traces = _build_network_docs(
        macro_raw, macro_grouped, NETWORK_META
    )

    subnetwork_docs: dict[str, Any] | None = None
    sub_summaries: dict[str, Any] | None = None
    sub_z_traces: dict[str, np.ndarray] | None = None

    if use_sub:
        sub_raw = compute_network_raw_traces(parcel_data, networks, macro=False)
        sub_grouped = _network_indices(networks, _subnetwork_order(networks))
        subnetwork_docs, sub_summaries, sub_z_traces, _ = _build_network_docs(
            sub_raw, sub_grouped, SUBNETWORK_META
        )

    z_stack = np.stack([z_traces[n] for n in YEO_NETWORK_ORDER], axis=0)
    dominant = [YEO_NETWORK_ORDER[i] for i in np.argmax(z_stack, axis=0).tolist()]

    salience_deriv = d_traces["SalVentAttn"]
    salience_events = [
        int(t)
        for t in np.where(salience_deriv >= SALIENCE_DERIVATIVE_THRESHOLD)[0].tolist()
        if t >= skip_trs
    ]

    coupling: dict[str, Any] = {}
    for a, b in COUPLING_PAIRS:
        key = f"{a}_{b}"
        series = _rolling_correlation(z_traces[a], z_traces[b], COUPLING_WINDOW_TRS)
        coupling[key] = {
            "networks": [a, b],
            "window_s": COUPLING_WINDOW_TRS,
            "series": series.tolist(),
        }

    segments = parse_dominant_segments(dominant)
    epoch_segments = classify_segment_epoch_templates(segments, z_traces, sub_z_traces)

    doc: dict[str, Any] = {
        "version": 2,
        "fps": fps,
        "n_trs": n_trs,
        "atlas": {
            "name": "Schaefer2018",
            "n_parcels": n_parcels,
            "yeo_networks": yeo_networks if use_sub else 7,
        },
        "inference_mode": inference_mode,
        "preprocessing": {
            "network_aggregate": "mean_over_parcels",
            "zscore": "per_network_within_clip",
            "skip_trs": skip_trs,
            "macro_rollup": use_sub,
        },
        "networks": network_docs,
        "summaries": summaries,
        "derived": {
            "dominant_network_tr": dominant,
            "dominant_segments": segments,
            "epoch_segments": epoch_segments,
            "coupling": coupling,
            "salience_events": {
                "threshold_z_derivative": SALIENCE_DERIVATIVE_THRESHOLD,
                "trs": salience_events,
            },
            "deep_dive_subnets": {
                macro: list(subs) for macro, subs in MACRO_DEEP_DIVE_SUBNETS.items()
            },
        },
        "disclaimer": DISCLAIMER,
    }

    if subnetwork_docs is not None:
        doc["subnetworks"] = subnetwork_docs
        doc["sub_summaries"] = sub_summaries

    return doc
